cores/esp32/IPAddress.h

```diff
@@ -1,105 +1,108 @@
 /*
- IPAddress.h - Base class that provides IPAddress
- Copyright (c) 2011 Adrian McEwen.  All right reserved.
+  IPAddress.h - Base class that provides IPAddress
+  Copyright (c) 2011 Adrian McEwen.  All right reserved.
 
- This library is free software; you can redistribute it and/or
- modify it under the terms of the GNU Lesser General Public
- License as published by the Free Software Foundation; either
- version 2.1 of the License, or (at your option) any later version.
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
 
- This library is distributed in the hope that it will be useful,
- but WITHOUT ANY WARRANTY; without even the implied warranty of
- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
- Lesser General Public License for more details.
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
 
- You should have received a copy of the GNU Lesser General Public
- License along with this library; if not, write to the Free Software
- Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
- */
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
 
-#ifndef IPAddress_h
-#define IPAddress_h
+#pragma once
 
 #include <stdint.h>
-#include <WString.h>
-#include <Printable.h>
-
-// A class to make it easier to handle and pass around IP addresses
+#include "Printable.h"
+#include "WString.h"
+#include "lwip/ip_addr.h"
 
 #define IPADDRESS_V4_BYTES_INDEX 12
 #define IPADDRESS_V4_DWORD_INDEX 3
 
-enum IPType
-{
+// A class to make it easier to handle and pass around IP addresses
+
+enum IPType {
     IPv4,
     IPv6
 };
 
-class IPAddress: public Printable
-{
+class IPAddress : public Printable {
 private:
     union {
         uint8_t bytes[16];
         uint32_t dword[4];
     } _address;
     IPType _type;
+    uint8_t _zone;
```

Comment:
Zone (or scope ID) is a bit of a tricky one, because it is part of the socket, but not part of the IPv6 address (although is part of, for example, a host literal address). See RFC 4007 and RFC 6874.

To be consistent with other implementations you might want to make these two separate classes: IPAddress and IPv6ScopeAddress.  IPAddress is as was previously (no scope field)

IPv6ScopeAddress has two fields: IPAddress and the Zone identifier.

In most cases (global addresses), zone identifier is not relevant.

Conversely, if you do want to combine Zone ID into IPAddress, then I suggest raising a discussion in ArduinoCore first, and see how they want to do it, so it is a common API across all Arduino usages. Breaking compatibility is not a good thing.

Adding support for Zone ID should also probably be a separate change. Just get basic IPv6 working first (global addresses). Smaller changes are easier to discuss, agree on, and merge.

(Unless you have a specific current scenario of needing link-local address scope identifiers. Maybe for Thread? Even then, I suggest making a separate PR. Small is better.)

Reply:
The question is not so much whether a zone id is part of an address or not, but how useful it is. A link-local address without its zone is more or less useless. If we don't include the zone within IPAddress, we need to duplicate many functions to accomodate for the specific case of link-local address with extra zone parameter. That would be counter-productive.

The main use for IPv6 we have currently is the Matter protocol that uses primarily link-local addresses. So it's not an exception by far.

I suppose that early implementation of Arduino-Core didn't have this problem if they had only one network interface.

Reply:
> Adding support for Zone ID should also probably be a separate change. Just get basic IPv6 working first (global addresses). Smaller changes are easier to discuss, agree on, and merge.

Not entirely sure if that's the best approach.
It seems to me it way easier to test locally first as that will eliminate lots of unknowns like poorly configured routers/firewalls etc.
I know IPv6 isn't something new (I have dual-stack at home since 2010), but I still see every now and then some funky stuff with IPv6-implementations on some hardware.

Also if Matter mostly uses IPv6 local link, then that's probably what people want to use IPv6 for right now.
Right now that's about the most sought after reason to implement IPv6 on microcontrollers.
I can't think of anything else which can't work without IPv6.

Reply:
Since zone is part of our network layer, we need to have support for it. This in no way interferes with the original implementation. It's OK to add functionality, as long as it does not break compatibility.

Reply:
I'm glad Matter+Thread requires IPv6; helps solve the chicken-and-egg (can't use IPv6 because no support; no support added because no one uses it).

Yes, link-local are useless without the zone identifier, so it makes sense to include them in the higher level abstraction (to most the distinction between an IPv6 address and a scoped IPv6 address is irrelevant). You might want to add some comments in the code and explain / refer to RFCs on what it is. (And maybe mention that IPAddress now represents a scoped address).

Overall, it is clear you have thought of the implications, and, for Arduino, which is supposed to simplify things, it makes sense to keep it all as one class.

One thing I was going to mention was to include them in printTo/toString and fromString, but it looks like you are already doing that.

Note that if you are talking about implementations using link-local addresses then it is absolutely clear that a simple hostByName6() would be quite useless as to whether it would return a global address or link-local (unlike IPv4, IPv6 normally has at least two addresses, and often more).  Something that returns multiple addresses is clearly needed (but I do suggest making it a separate pull request if possible).

Reply:
Agreed, thanks.